### PR TITLE
os/bluestore/NVMEDevice: fix the nvme queue depth issue

### PR DESCRIPTION
**Pull request by optimistyzy**
We need to track the queue depth usage for each qpair to
prevent submit too many I/Os. If there are too many requests,
we need to poll the qpair first in order to make sure there are enough slots
for next I/O submit.

Signed-off-by: Ziye Yang <optimistyzy@gmail.com>
Signed-off-by: Pan Liu <wanjun.lp@alibaba-inc.com>

### CODE REVIEW

**Comment by optimistyzy** [comment]
@liupan1111  @yuyuyu101

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -138,8 +142,14 @@ class SharedDriverQueueData {
         flush_lock("NVMEDevice::flush_lock"),
         flush_waiters(0),
         completed_op_seq(0), queue_op_seq(0) {
+    struct spdk_nvme_io_qpair_opts opts = {};
```
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: indent issue?

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -138,8 +142,14 @@ class SharedDriverQueueData {
...
+    spdk_nvme_ctrlr_get_default_io_qpair_opts(ctrlr, &opts, sizeof(opts));
+    opts.qprio = SPDK_NVME_QPRIO_URGENT;
+    opts.io_queue_size = std::min(queue_depth, opts.io_queue_size);
+    max_queue_depth = opts.io_queue_size - 1;
```
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: why io_queue_size - 1? thanks.

File: src/os/bluestore/NVMEDevice.cc [hunk 3]
```diff
@@ -470,6 +480,10 @@ void SharedDriverQueueData::_aio_thread()
     }
 
     for (; t; t = t->next) {
+      // no slots
+      if (current_queue_depth == max_queue_depth)
```
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: move "// no slots" here?

File: src/os/bluestore/NVMEDevice.cc [hunk 4]
```diff
@@ -138,8 +142,14 @@ class SharedDriverQueueData {
         flush_lock("NVMEDevice::flush_lock"),
         flush_waiters(0),
         completed_op_seq(0), queue_op_seq(0) {
+    struct spdk_nvme_io_qpair_opts opts = {};
+    spdk_nvme_ctrlr_get_default_io_qpair_opts(ctrlr, &opts, sizeof(opts));
```
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: @optimistyzy , please correct me: 
this function is used to get io_queue_size(default queue pair iodepth), so why not just use this default value as max_queue_depth?

File: src/os/bluestore/NVMEDevice.cc [hunk 5]
```diff
@@ -539,6 +553,7 @@ void SharedDriverQueueData::_aio_thread()
           break;
         }
       }
+      current_queue_depth++;
```
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: should this also supported in sync write and read?

File: src/os/bluestore/NVMEDevice.cc [hunk 6]
```diff
@@ -801,6 +816,7 @@ void io_complete(void *t, const struct spdk_nvme_cpl *completion)
   assert(queue != NULL);
   assert(ctx != NULL);
   ++queue->completed_op_seq;
+  --queue->current_queue_depth;
```
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: 1）should this also supported in sync write and read?
2）should it be defined as atomic_t?

**Comment by liupan1111** [comment]
@optimistyzy As I understand, this PR will prevent from "ceph_abort" when submitting too many IOs, but will not related to the problem IOPS drops to zero after several mins. Am I right?

File: src/os/bluestore/NVMEDevice.cc (hunk 2 above)
Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: in spdk program，if user want to queue depth =n，we should set n+1

File: src/os/bluestore/NVMEDevice.cc (hunk 4 above)
Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: this is used to get hardware supported size，but not the driver supported

File: src/os/bluestore/NVMEDevice.cc (hunk 6 above)
Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: not needed，it is only operated buy one thread

**Comment by optimistyzy** [comment]
it is related with drop to zero issue

File: src/os/bluestore/NVMEDevice.cc (hunk 2 above)
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: ok... It is tricky here, suggest we add a comment in the code...

File: src/os/bluestore/NVMEDevice.cc (hunk 4 above)
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: do you mean spdk userspace nvme driver? If yes, isn't it supply a function to get the size driver supported?

Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: As we talked in WeChat, it is used to get hardware info, spdk doesn't support get default driver depth.

File: src/os/bluestore/NVMEDevice.cc (hunk 6 above)
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: I understand now... because of the spike lib limit, submit and poll should be in the same thread...

**Comment by liupan1111** [comment]
I've tested it locally, by P3500 intel ssd, it works find.

**Comment by liupan1111** [comment]
@yuyuyu101 what's your opinion?

File: src/os/bluestore/NVMEDevice.cc [hunk 7]
```diff
@@ -470,6 +478,11 @@ void SharedDriverQueueData::_aio_thread()
     }
 
     for (; t; t = t->next) {
+      if (current_queue_depth == max_queue_depth) {
+        // no slots
+        goto again;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: I think this is a little rough. do we need to introduce backoff method to avoid cpu spining?

Comment (liupan1111) [reply src/os/bluestore/NVMEDevice.cc]: yes, I think so ... similar as what aio.cc does... @optimistyzy could you modify it?

**Comment by optimistyzy** [comment]
@yuyuyu101  @liupan1111  I do not follow up. In my mind,  we should call spdk_nvme_qpair_process_completions (jump to try again), otherwise those already submitted I/O will not be finished (call the io_complete), so sleep is not helpful.  See spdk's nvme perf tool, it also does the logic in this way.

File: src/os/bluestore/NVMEDevice.cc [hunk 8]
```diff
@@ -118,6 +119,7 @@ class SharedDriverQueueData {
   std::set<uint64_t> flush_waiter_seqs;
 
   public:
+    uint32_t current_queue_depth;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: current_queue_depth = 0

Comment (optimistyzy) [reply src/os/bluestore/NVMEDevice.cc]: the default value is zero？

**Pull request merged**